src/Components/WebAssembly/DebugProxy/src/TargetPickerUi.cs

```diff
@@ -0,0 +1,223 @@
+            IgnoreNullValues = true
+        };
+
+        private readonly DebugProxyOptions _options;
+
+        public TargetPickerUi(DebugProxyOptions options)
+        {
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        public async Task Display(HttpContext context)
+        {
+            context.Response.ContentType = "text/html";
+
+            var request = context.Request;
+            var targetApplicationUrl = request.Query["url"];
+
+            var debuggerTabsListUrl = $"{_options.BrowserHost}/json";
+            IEnumerable<BrowserTab> availableTabs;
+
+            try
+            {
+                availableTabs = await GetOpenedBrowserTabs();
+            }
+            catch (Exception ex)
+            {
+                await context.Response.WriteAsync($@"
+<h1>Unable to find debuggable browser tab</h1>
+<p>
+    Could not get a list of browser tabs from <code>{debuggerTabsListUrl}</code>.
+    Ensure your browser is running with debugging enabled.
+</p>
+<h2>Resolution</h2>
+<p>
+    <h4>If you are using Google Chrome for your development, follow these instructions:</h4>
+    {GetLaunchChromeInstructions(targetApplicationUrl)}
+</p>
+<p>
+    <h4>If you are using Microsoft Edge (Chromium) for your development, follow these instructions:</h4>
+    {GetLaunchEdgeInstructions(targetApplicationUrl)}
+</p>
+<strong>This should launch a new browser window with debugging enabled..</p>
+<h2>Underlying exception:</h2>
+<pre>{ex}</pre>
+                ");
+
+                return;
+            }
+
+            var matchingTabs = string.IsNullOrEmpty(targetApplicationUrl)
+                ? availableTabs.ToList()
+                : availableTabs.Where(t => t.Url.Equals(targetApplicationUrl, StringComparison.Ordinal)).ToList();
+
+            if (matchingTabs.Count == 1)
+            {
+                // We know uniquely which tab to debug, so just redirect
+                var devToolsUrlWithProxy = GetDevToolsUrlWithProxy(request, matchingTabs.Single());
+                context.Response.Redirect(devToolsUrlWithProxy);
+            }
+            else if (matchingTabs.Count == 0)
+            {
+                await context.Response.WriteAsync("<h1>No inspectable pages found</h1>");
+
+                var suffix = string.IsNullOrEmpty(targetApplicationUrl)
+                    ? string.Empty
+                    : $" matching the URL {WebUtility.HtmlEncode(targetApplicationUrl)}";
+                await context.Response.WriteAsync($"<p>The list of targets returned by {WebUtility.HtmlEncode(debuggerTabsListUrl)} contains no entries{suffix}.</p>");
+                await context.Response.WriteAsync("<p>Make sure your browser is displaying the target application.</p>");
+            }
+            else
+            {
+                await context.Response.WriteAsync("<h1>Inspectable pages</h1>");
+                await context.Response.WriteAsync(@"
```

Comment:
If only we had a UI stack for rendering HTML. :trollface:

Reply:
I can't think of any situation where anyone would need that. Or more than 640Kb of RAM.